File: security-platform/agents/go/middleware.go
package agent

import (
	"net/http"

	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/trace"
)

// HTTPMiddleware creates HTTP middleware for automatic instrumentation.
func (a *Agent) HTTPMiddleware(next http.Handler) http.Handler {
	return otelhttp.NewHandler(
		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ctx := r.Context()
			span := trace.SpanFromContext(ctx)

			// Add request attributes
			span.SetAttributes(
				attribute.String("http.method", r.Method),
				attribute.String("http.url", r.URL.String()),
				attribute.String("http.user_agent", r.UserAgent()),
			)

			// Check policy if in block mode
			if a.config.Policy != nil && a.config.Policy.Mode == "block" {
				result := a.CheckPolicy("http.request", map[string]interface{}{
					"method": r.Method,
					"path":   r.URL.Path,
					"ip":     r.RemoteAddr,
				})

				if !result.Allowed {
					span.SetAttributes(attribute.Bool("policy.blocked", true))
					span.SetStatus(codes.Error, result.Reason)
					http.Error(w, "Request blocked by policy", http.StatusForbidden)
					return
				}
			}

			// Call next handler
			next.ServeHTTP(w, r.WithContext(ctx))
		}),
		"http.request",
		otelhttp.WithTracerProvider(otel.GetTracerProvider()),
	)
}

File: security-platform/agents/node/src/middleware.ts
import { Request, Response, NextFunction } from 'express';
import { Agent } from './agent';
import { MiddlewareOptions } from './types';

export function createMiddleware(agent: Agent, options: MiddlewareOptions = {}) {
  return (req: Request, res: Response, next: NextFunction) => {
    const startTime = Date.now();

    // Capture request metadata
    const requestMetadata = {
      method: req.method,
      path: req.path,
      headers: agent.redact(req.headers),
      query: agent.redact(req.query),
      body: agent.redact(req.body),
      ip: req.ip,
      userAgent: req.get('user-agent'),
    };

    // Check policy
    const policyResult = agent.checkPolicy('http_request', {
      method: req.method,
      path: req.path,
      ip: req.ip,
    });

    if (!policyResult.allowed && options.mode === 'block') {
      res.status(403).json({ error: 'Request blocked by security policy', reason: policyResult.reason });
      return;
    }

    // Capture response
    const originalSend = res.send;
    res.send = function (body: any) {
      const responseMetadata = {
        statusCode: res.statusCode,
        headers: agent.redact(res.getHeaders()),
        body: options.captureResponseBody ? agent.redact(body) : undefined,
        duration: Date.now() - startTime,
      };

      // Send to control plane (async, non-blocking)
      if (options.sendToControlPlane !== false) {
        sendEvent('http_request', {
          request: requestMetadata,
          response: responseMetadata,
          policy: policyResult,
        }, agent).catch(err => {
          console.error('Failed to send event:', err);
        });
      }

      return originalSend.call(this, body);
    };

    next();
  };
}

async function sendEvent(type: string, data: any, agent: Agent): Promise<void> {
  // In real implementation, send to control plane via HTTP or OTLP
  // For now, this is a placeholder
}
